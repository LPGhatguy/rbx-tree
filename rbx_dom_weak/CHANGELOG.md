# rbx\_dom\_weak Changelog

## Unreleased
* `RbxValueType` now derives `Serialize` and `Deserialize`
* Added support for `BrickColor`

## 1.4.0 (2019-03-27)
* Added support for new types:
	* `NumberRange`
	* [#15](https://github.com/LPGhatguy/rbx-dom/pull/15):
		* `Int64`
		* `Float64`
	* [#20](https://github.com/LPGhatguy/rbx-dom/pull/20):
		* `Rect`
		* `Ray`
		* `ColorSequence`
		* `NumberSequence`

## 1.3.0 (2019-03-14)
* Added `sort_children_by_key` and `sort_children_unstable_by_key` methods to `RbxInstance` to reorder children safely

## 1.2.0 (2019-03-13)
* `RbxTree` and `RbxInstance` are now clonable without an `unimplemented` panic
	* Cloned trees preserve their IDs as-is and there's no public API yet to transplant instances between trees. This is mostly useful for comparing trees before/after a mutation, which we're using in Rojo.

## 1.1.0 (2019-03-11)
* Marked `PhysicalProperties` as `pub`
* Fixed `Serialize` impl for `UnresolvedRbxValue`
	* This removes `Serialize` from `AmbiguousRbxValue`, which should not break any real code.

## 1.0.0 (2019-03-01)
* `RbxValue` and `RbxValueType` can no longer be matched exhaustively, which enables adding new types without breaking code in the future
* Added support for `UDim`, `UDim2`, and `Content` value types
* `PhysicalProperties` is no longer a stub type

## 0.3.1 (2019-02-26)
* Added support for bare bool values when deserializing `UnresolvedRbxValue`
* Implemented `Serialize` for `UnresolvedRbxValue` and `AmbiguousRbxValue`
* Implemented `From<RbxValue>` for `UnresolvedRbxValue`

## 0.3.0 (2019-02-14)
* Renamed crate from `rbx_tree` to `rbx_dom_weak`
* Added support for `Ref` values ([#8](https://github.com/LPGhatguy/rbx-dom/pull/8))
* Added `UnresolvedRbxValue` and `AmbiguousRbxValue`, intended to be used alongside `rbx_reflection` to make specifying values less verbose.

## 0.2.0 (2019-01-25)
* Added new variants for `RbxValue`:
	* Int32
	* Float32
	* Enum
	* Vector2
	* Color3
	* Color3uint8
	* Vector3int16
	* Vector2int16
	* CFrame
	* PhysicalProperties (Stub)

## 0.1.0
* Initial release
* Supports `String`, `Bool`, and `Vector3` property values