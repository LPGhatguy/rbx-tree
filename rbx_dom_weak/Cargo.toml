[package]
name = "rbx_dom_weak"
description = "Weakly-typed Roblox DOM implementation for Rust"
license = "MIT"
documentation = "https://docs.rs/rbx_dom_weak"
homepage = "https://github.com/LPGhatguy/rbx-dom"
repository = "https://github.com/LPGhatguy/rbx-dom.git"
version = "1.8.2"
authors = ["Lucien Greathouse <me@lpghatguy.com>"]
edition = "2018"

[dependencies]
base64 = "0.10.1"
lazy_static = "1.3.0"
md5 = "0.6.1"
serde = "1.0"
serde_derive = "1.0"
serde_repr = "0.1"
uuid = { version = "0.7", features = ["v4", "serde"] }

[dev-dependencies]
serde_json = "1.0"
maplit = "1.0"
