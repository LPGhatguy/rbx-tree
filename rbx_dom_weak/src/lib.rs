//! rbx_dom_weak is a common representation of the Roblox DOM for Rust. It's
//! designed to play nicely with the borrow checker and allows accessing
//! instances by ID in constant time.
//!
//! Constructing a new tree of instances is accomplished by first creating an
//! [`RbxInstanceProperties`] object that describes the root instance of the
//! tree, and then wrapping it with an [`RbxTree`]:
//!
//! ```
//! use std::collections::HashMap;
//! use rbx_dom_weak::{RbxInstanceProperties, RbxTree};
//!
//! let props = RbxInstanceProperties {
//!     name: "My Cool Game".to_owned(),
//!     class_name: "DataModel".to_owned(),
//!     properties: HashMap::new(),
//! };
//!
//! let mut tree = RbxTree::new(props);
//! println!("ID of instance we just inserted is {}", tree.get_root_id());
//! ```
//!
//! Note that the [maplit] crate is incredibly useful for defining properties
//! inline.
//!
//! Once we have a tree, we can use [`RbxTree::insert_instance`] and
//! [`RbxTree::get_instance`] to add instances to the tree and retrieve them.
//!
//! ```
//! # use std::collections::HashMap;
//! # use rbx_dom_weak::{RbxInstanceProperties, RbxTree};
//! use rbx_dom_weak::RbxValue;
//! use maplit::hashmap;
//! #
//! # let props = RbxInstanceProperties {
//! #     name: "My Cool Game".to_owned(),
//! #     class_name: "DataModel".to_owned(),
//! #     properties: HashMap::new(),
//! # };
//! #
//! # let mut tree = RbxTree::new(props);
//! #
//! let http_service = RbxInstanceProperties {
//!     name: "HttpService".to_owned(),
//!     class_name: "HttpService".to_owned(),
//!     properties: hashmap! {
//          // Properties are represented via the RbxValue enum
//!         "HttpEnabled".to_owned() => RbxValue::Bool {
//!             value: true,
//!         },
//!     },
//! };
//!
//! let datamodel_id = tree.get_root_id();
//! let http_service_id = tree.insert_instance(http_service, datamodel_id);
//!
//! println!("HttpService has ID {}", http_service_id);
//! ```
//!
//! To change properties on an instance that's already present in the tree, use
//! [`RbxTree::get_instance_mut`]. Note that it isn't possible to add or remove
//! children through this method, use [`RbxTree::insert_instance`] instead.
//!
//! [`RbxTree`]: struct.RbxTree.html
//! [`RbxTree::insert_instance`]: struct.RbxTree.html#method.insert_instance
//! [`RbxTree::get_instance`]: struct.RbxTree.html#method.get_instance
//! [`RbxTree::get_instance_mut`]: struct.RbxTree.html#method.get_instance_mut
//! [`RbxInstanceProperties`]: struct.RbxInstanceProperties.html
//! [maplit]: https://crates.io/crates/maplit

mod unresolved_value;
mod brick_color;
mod id;
mod instance;
mod tree;
mod value;
mod value_conversion;

pub use crate::{
    unresolved_value::{AmbiguousRbxValue, UnresolvedRbxValue},
    brick_color::BrickColor,
    id::RbxId,
    instance::{RbxInstanceProperties, RbxInstance},
    tree::{RbxTree, Descendants},
    value::*,
};