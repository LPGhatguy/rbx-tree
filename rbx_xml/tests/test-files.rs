use std::{fs, path::Path};

use rbx_dom_weak::DomViewer;

macro_rules! test_models {
    ( $( $test_name: ident : $file_name: expr,)* ) => {
        $(
            #[test]
            fn $test_name() {
                let _ = env_logger::try_init();

                let mut path = Path::new("../test-files/models").join($file_name);
                path.push("xml.rbxmx");

                let contents = fs::read_to_string(path).unwrap();
                let dom = rbx_xml::from_str_default(&contents).unwrap();

                let mut viewer = DomViewer::new();

                insta::assert_yaml_snapshot!(viewer.view_children(&dom));
            }
        )*
    };
}

test_models! {
    default_inserted_folder: "default-inserted-folder",
    default_inserted_part: "default-inserted-part",
    // faces: "faces",
    // axes: "axes",
    ref_adjacent: "ref-adjacent",
    ref_child: "ref-child",
    ref_parent: "ref-parent",
    body_movers: "body-movers",
    // default_inserted_modulescript: "default-inserted-modulescript",
}
