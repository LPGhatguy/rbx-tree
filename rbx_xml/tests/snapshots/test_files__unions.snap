---
source: rbx_xml/tests/test-files.rs
expression: viewer.view_children(&dom)
---
- referent: referent-0
  name: Red Union 1
  class: UnionOperation
  properties:
    Anchored:
      Type: Bool
      Value: false
    AttributesSerialize:
      Type: BinaryString
      Value: ""
    BackParamA:
      Type: Float32
      Value: -0.5
    BackParamB:
      Type: Float32
      Value: 0.5
    BackSurface:
      Type: EnumValue
      Value: 0
    BackSurfaceInput:
      Type: EnumValue
      Value: 0
    BottomParamA:
      Type: Float32
      Value: -0.5
    BottomParamB:
      Type: Float32
      Value: 0.5
    BottomSurface:
      Type: EnumValue
      Value: 0
    BottomSurfaceInput:
      Type: EnumValue
      Value: 0
    CFrame:
      Type: CFrame
      Value:
        Position:
          - 0.0
          - 0.0
          - 0.0
        Orientation:
          - - 1.0
            - 0.0
            - 0.0
          - - 0.0
            - 1.0
            - 0.0
          - - 0.0
            - 0.0
            - 1.0
    CanCollide:
      Type: Bool
      Value: true
    CastShadow:
      Type: Bool
      Value: true
    CollisionGroupId:
      Type: Int32
      Value: 0
    Color:
      Type: Color3uint8
      Value:
        - 196
        - 40
        - 28
    CustomPhysicalProperties:
      Type: PhysicalProperties
      Value: Default
    FrontParamA:
      Type: Float32
      Value: -0.5
    FrontParamB:
      Type: Float32
      Value: 0.5
    FrontSurface:
      Type: EnumValue
      Value: 0
    FrontSurfaceInput:
      Type: EnumValue
      Value: 0
    LeftParamA:
      Type: Float32
      Value: -0.5
    LeftParamB:
      Type: Float32
      Value: 0.5
    LeftSurface:
      Type: EnumValue
      Value: 0
    LeftSurfaceInput:
      Type: EnumValue
      Value: 0
    Locked:
      Type: Bool
      Value: false
    Massless:
      Type: Bool
      Value: false
    Material:
      Type: EnumValue
      Value: 256
    PhysicalConfigData:
      len: 1803
      hash: db0e2bce2a871addfc49da3894a9b0c8ef97b3b83c962717ece20aa97635fee3
    Reflectance:
      Type: Float32
      Value: 0.0
    RenderFidelity:
      Type: EnumValue
      Value: 1
    RightParamA:
      Type: Float32
      Value: -0.5
    RightParamB:
      Type: Float32
      Value: 0.5
    RightSurface:
      Type: EnumValue
      Value: 0
    RightSurfaceInput:
      Type: EnumValue
      Value: 0
    RootPriority:
      Type: Int32
      Value: 0
    RotVelocity:
      Type: Vector3
      Value:
        - 0.0
        - 0.0
        - 0.0
    Size:
      Type: Vector3
      Value:
        - 5.0
        - 2.0
        - 2.0
    SmoothingAngle:
      Type: Float32
      Value: 0.0
    SourceAssetId:
      Type: Int64
      Value: -1
    Tags:
      Type: BinaryString
      Value: ""
    TopParamA:
      Type: Float32
      Value: -0.5
    TopParamB:
      Type: Float32
      Value: 0.5
    TopSurface:
      Type: EnumValue
      Value: 0
    TopSurfaceInput:
      Type: EnumValue
      Value: 0
    Transparency:
      Type: Float32
      Value: 0.0
    UsePartColor:
      Type: Bool
      Value: false
    Velocity:
      Type: Vector3
      Value:
        - 0.0
        - 0.0
        - 0.0
  children: []
- referent: referent-1
  name: Red Union 2
  class: UnionOperation
  properties:
    Anchored:
      Type: Bool
      Value: false
    AttributesSerialize:
      Type: BinaryString
      Value: ""
    BackParamA:
      Type: Float32
      Value: -0.5
    BackParamB:
      Type: Float32
      Value: 0.5
    BackSurface:
      Type: EnumValue
      Value: 0
    BackSurfaceInput:
      Type: EnumValue
      Value: 0
    BottomParamA:
      Type: Float32
      Value: -0.5
    BottomParamB:
      Type: Float32
      Value: 0.5
    BottomSurface:
      Type: EnumValue
      Value: 0
    BottomSurfaceInput:
      Type: EnumValue
      Value: 0
    CFrame:
      Type: CFrame
      Value:
        Position:
          - -7.0
          - 0.0
          - 0.0
        Orientation:
          - - 1.0
            - 0.0
            - 0.0
          - - 0.0
            - 1.0
            - 0.0
          - - 0.0
            - 0.0
            - 1.0
    CanCollide:
      Type: Bool
      Value: true
    CastShadow:
      Type: Bool
      Value: true
    CollisionGroupId:
      Type: Int32
      Value: 0
    Color:
      Type: Color3uint8
      Value:
        - 196
        - 40
        - 28
    CustomPhysicalProperties:
      Type: PhysicalProperties
      Value: Default
    FrontParamA:
      Type: Float32
      Value: -0.5
    FrontParamB:
      Type: Float32
      Value: 0.5
    FrontSurface:
      Type: EnumValue
      Value: 0
    FrontSurfaceInput:
      Type: EnumValue
      Value: 0
    LeftParamA:
      Type: Float32
      Value: -0.5
    LeftParamB:
      Type: Float32
      Value: 0.5
    LeftSurface:
      Type: EnumValue
      Value: 0
    LeftSurfaceInput:
      Type: EnumValue
      Value: 0
    Locked:
      Type: Bool
      Value: false
    Massless:
      Type: Bool
      Value: false
    Material:
      Type: EnumValue
      Value: 256
    PhysicalConfigData:
      len: 1803
      hash: db0e2bce2a871addfc49da3894a9b0c8ef97b3b83c962717ece20aa97635fee3
    Reflectance:
      Type: Float32
      Value: 0.0
    RenderFidelity:
      Type: EnumValue
      Value: 1
    RightParamA:
      Type: Float32
      Value: -0.5
    RightParamB:
      Type: Float32
      Value: 0.5
    RightSurface:
      Type: EnumValue
      Value: 0
    RightSurfaceInput:
      Type: EnumValue
      Value: 0
    RootPriority:
      Type: Int32
      Value: 0
    RotVelocity:
      Type: Vector3
      Value:
        - 0.0
        - 0.0
        - 0.0
    Size:
      Type: Vector3
      Value:
        - 5.0
        - 2.0
        - 2.0
    SmoothingAngle:
      Type: Float32
      Value: 0.0
    SourceAssetId:
      Type: Int64
      Value: -1
    Tags:
      Type: BinaryString
      Value: ""
    TopParamA:
      Type: Float32
      Value: -0.5
    TopParamB:
      Type: Float32
      Value: 0.5
    TopSurface:
      Type: EnumValue
      Value: 0
    TopSurfaceInput:
      Type: EnumValue
      Value: 0
    Transparency:
      Type: Float32
      Value: 0.0
    UsePartColor:
      Type: Bool
      Value: false
    Velocity:
      Type: Vector3
      Value:
        - 0.0
        - 0.0
        - 0.0
  children: []
- referent: referent-2
  name: Blue Union
  class: UnionOperation
  properties:
    Anchored:
      Type: Bool
      Value: false
    AttributesSerialize:
      Type: BinaryString
      Value: ""
    BackParamA:
      Type: Float32
      Value: -0.5
    BackParamB:
      Type: Float32
      Value: 0.5
    BackSurface:
      Type: EnumValue
      Value: 0
    BackSurfaceInput:
      Type: EnumValue
      Value: 0
    BottomParamA:
      Type: Float32
      Value: -0.5
    BottomParamB:
      Type: Float32
      Value: 0.5
    BottomSurface:
      Type: EnumValue
      Value: 0
    BottomSurfaceInput:
      Type: EnumValue
      Value: 0
    CFrame:
      Type: CFrame
      Value:
        Position:
          - -4.0
          - -3.5
          - 0.0
        Orientation:
          - - 1.0
            - 0.0
            - 0.0
          - - 0.0
            - 1.0
            - 0.0
          - - 0.0
            - 0.0
            - 1.0
    CanCollide:
      Type: Bool
      Value: true
    CastShadow:
      Type: Bool
      Value: true
    CollisionGroupId:
      Type: Int32
      Value: 0
    Color:
      Type: Color3uint8
      Value:
        - 13
        - 105
        - 172
    CustomPhysicalProperties:
      Type: PhysicalProperties
      Value: Default
    FrontParamA:
      Type: Float32
      Value: -0.5
    FrontParamB:
      Type: Float32
      Value: 0.5
    FrontSurface:
      Type: EnumValue
      Value: 0
    FrontSurfaceInput:
      Type: EnumValue
      Value: 0
    LeftParamA:
      Type: Float32
      Value: -0.5
    LeftParamB:
      Type: Float32
      Value: 0.5
    LeftSurface:
      Type: EnumValue
      Value: 0
    LeftSurfaceInput:
      Type: EnumValue
      Value: 0
    Locked:
      Type: Bool
      Value: false
    Massless:
      Type: Bool
      Value: false
    Material:
      Type: EnumValue
      Value: 256
    PhysicalConfigData:
      len: 1367
      hash: cc62d7045ec238081b7ebbd97ce5ce7dc15166e688168c307247089135718bdb
    Reflectance:
      Type: Float32
      Value: 0.0
    RenderFidelity:
      Type: EnumValue
      Value: 1
    RightParamA:
      Type: Float32
      Value: -0.5
    RightParamB:
      Type: Float32
      Value: 0.5
    RightSurface:
      Type: EnumValue
      Value: 0
    RightSurfaceInput:
      Type: EnumValue
      Value: 0
    RootPriority:
      Type: Int32
      Value: 0
    RotVelocity:
      Type: Vector3
      Value:
        - 0.0
        - 0.0
        - 0.0
    Size:
      Type: Vector3
      Value:
        - 4.0
        - 2.0
        - 2.0
    SmoothingAngle:
      Type: Float32
      Value: 0.0
    SourceAssetId:
      Type: Int64
      Value: -1
    Tags:
      Type: BinaryString
      Value: ""
    TopParamA:
      Type: Float32
      Value: -0.5
    TopParamB:
      Type: Float32
      Value: 0.5
    TopSurface:
      Type: EnumValue
      Value: 0
    TopSurfaceInput:
      Type: EnumValue
      Value: 0
    Transparency:
      Type: Float32
      Value: 0.0
    UsePartColor:
      Type: Bool
      Value: false
    Velocity:
      Type: Vector3
      Value:
        - 0.0
        - 0.0
        - 0.0
  children: []
